### PR TITLE
Fix XDP Cleanup issue (#2619)

### PR DESCRIPTION
**Pull request by sjh5205**
This patch add sort of validation logic in RemoveXDP() function.
Before remove XDP program from target interface, it checks 'name' in currently loaded BPF program.
If it doesn't match with rule, (for example, 'prefilter' in bpf/bpf-apache/filter.c)
validation logic trigger error and XDP resync will be fail.

- bpf: add 'name' attribute in progInfo struct
- bpf: add getCalicoBPFProgIds()
- bpf: add canRemoveXDP()
- bpf: modify return value in GetXDPID() to know why this func is failed
- bpf: modify RemoveXDP()

It tested on my system, works well as I intended.
Also, because of recently added bpf-gpl/xdp.c, I think we should have common prefix to function name.

Thank you!

Known issue:
- validation routine prints the same error log multiple times

## Todos
- [ ] Unit tests (full coverage)
- [ ] Integration tests (delete as appropriate) In plan/Not needed/Done
- [ ] Documentation
- [ ] Backport
- [ ] Release note

## Release Note

```release-note
None required
```

### CODE REVIEW

**Comment by sjh5205** [comment]
2021-07-17 06:44:23.156 [INFO][72] felix/xdp_state.go 192: Applying BPF actions did not succeed. Queueing XDP resync. 

> error=failed to remove XDP program from enp137s0: [failed to detach XDP program from enp137s0: interface has bogus, but not loaded by Calico. failed to detach XDP program from enp137s0: interface has bogus, but not loaded by Calico. failed to detach XDP program from enp137s0: interface has bogus, but not loaded by Calico.]
> 2021-07-17 06:44:23.156 [WARNING][72] felix/int_dataplane.go 1273: failed to wipe the XDP state error=failed to remove XDP program from enp137s0: [failed to detach XDP program from enp137s0: interface has bogus, but not loaded by Calico. failed to detach XDP program from enp137s0: interface has bogus, but not loaded by Calico. failed to detach XDP program from enp137s0: interface has bogus, but not loaded by Calico.] try=5
> 2021-07-17 06:44:23.439 [INFO][72] felix/xdp_state.go 561: Finished XDP resync. family=4 resyncDuration=183.356329ms
> 2021-07-17 06:44:23.452 [INFO][72] felix/xdp_state.go 192: Applying BPF actions did not succeed. Queueing XDP resync. error=failed to remove XDP program from enp137s0: [failed to detach XDP program from enp137s0: interface has bogus, but not loaded by Calico. failed to detach XDP program from enp137s0: interface has bogus, but not loaded by Calico. failed to detach XDP program from enp137s0: interface has bogus, but not loaded by Calico.]
> 2021-07-17 06:44:23.452 [WARNING][72] felix/int_dataplane.go 1273: failed to wipe the XDP state error=failed to remove XDP program from enp137s0: [failed to detach XDP program from enp137s0: interface has bogus, but not loaded by Calico. failed to detach XDP program from enp137s0: interface has bogus, but not loaded by Calico. failed to detach XDP program from enp137s0: interface has bogus, but not loaded by Calico.] try=6
> 2021-07-17 06:44:26.531 [INFO][72] felix/xdp_state.go 561: Finished XDP resync. family=4 resyncDuration=2.978844402s

Here's example of the error log.
It prints same error message multiple times.

File: bpf/bpf.go [hunk 1]
```diff
@@ -1024,6 +1025,59 @@ func (b *BPFLib) getMapArgs(ifName string) ([]string, error) {
...
+	}
+
+	return id, nil
+}
+
+func (b *BPFLib) canRemoveXDP(ifName string, id []int) (bool, error) {
+	i, err := b.GetXDPID(ifName)
+	if err != nil {
+		// iproute2 error or Empty
+		if i == -1 {
```
Comment (neiljerram) [inline bpf/bpf.go:1066]: Please comment here to say what this case means.

File: bpf/bpf.go [hunk 2]
```diff
@@ -1024,6 +1025,59 @@ func (b *BPFLib) getMapArgs(ifName string) ([]string, error) {
...
+			return false, err
+		} else {
```
Comment (neiljerram) [inline bpf/bpf.go:1068]: And this one.

File: bpf/bpf.go [hunk 3]
```diff
@@ -1041,6 +1095,19 @@ func (b *BPFLib) RemoveXDP(ifName string, mode XDPMode) error {
 	progName := getProgName(ifName)
 	progPath := filepath.Join(b.xdpDir, progName)
 
+	// Verify Calico XDP program
+	id, err := b.getCalicoBPFProgIds()
+	if err != nil {
+		return err
+	}
+	verify, err := b.canRemoveXDP(ifName, id)
```
Comment (neiljerram) [inline bpf/bpf.go:1103]: I'm unclear what the return bool value of `canRemoveXDP` means, but I think it is "There is an XDP program and it's Calico, so please remove it"  Is that right?

File: bpf/bpf.go [hunk 4]
```diff
@@ -1210,7 +1277,7 @@ func (b *BPFLib) GetXDPID(ifName string) (int, error) {
 		}
 	}
 
-	return -1, errors.New("ID not found")
+	return -2, errors.New("ID not found")
```
Comment (neiljerram) [inline bpf/bpf.go:1280]: Why this change?

**Review by neiljerram — CHANGES_REQUESTED** [review]
Not sure I fully understand this yet; have dropped a few comments to help clarify.

**Comment by caseydavenport** [comment]
@sjh5205 ping - still working on this one?

**Comment by sjh5205** [comment]
@caseydavenport Oh, sorry. I lost notification mail in my mailbox.
I'll check the comments, then reply soon.

File: bpf/bpf.go (hunk 1 above)
Comment (sjh5205) [reply bpf/bpf.go:1066]: Originally, GetXDPID() returns -1 regardless of the type of error. But in my implementation I need to know why iproute2 returned an error.
This is because, even if GetXDPID() does not find the id (=even if the XDP program is not loaded), the RemoveXDP() function needs to be executed in the resync period.

I also think this method is not neat. Maybe it would be better to separate a function that checks for the existence of an XDP program.
I want to hear from you.

File: bpf/bpf.go (hunk 2 above)
Comment (sjh5205) [reply bpf/bpf.go:1068]: -1: Failed to get interface information
-2: Failed to get XDP prog ID, maybe just XDP program is not loaded yet.

-1 is clearly a situation we don't want, but in case of -2, we have to proceed RemoveXDP().

File: bpf/bpf.go (hunk 3 above)
Comment (sjh5205) [reply bpf/bpf.go:1103]: Basically, yes.

File: bpf/bpf.go (hunk 4 above)
Comment (sjh5205) [reply bpf/bpf.go:1280]: Answered above.

File: bpf/bpf.go (hunk 1 above)
Comment (neiljerram) [reply bpf/bpf.go:1066]: Sorry, still understanding some detail here.  Why is `RemoveXDP()` needed if there isn't any XDP program attached to the interface?

File: bpf/bpf.go (hunk 2 above)
Comment (neiljerram) [reply bpf/bpf.go:1068]: It will probably be better to encode the type of error in the Error object - but I'd like to make sure I understand the overall picture first.

File: bpf/bpf.go (hunk 3 above)
Comment (neiljerram) [reply bpf/bpf.go:1103]: Can you add a comment to say that above the definition of `canRemoveXDP`?

File: bpf/bpf.go (hunk 1 above)
Comment (sjh5205) [reply bpf/bpf.go:1066]: Ah, now I understand what you said.
If the interface does not have an XDP program, fixupXDPProgramAndMapConsistency() does not add UninstallXDP, and RemoveXDP() is not executed in XDPAction.apply().

Therefore, it is unnecessary to check whether the XDP program exists in GetXDPID().
Am I right?

**Comment by caseydavenport** [comment]
The Calico team is reorganizing our code and repository structure in order to make development easier, and as such all of the code in this repository is now located at https://github.com/projectcalico/calico. We appreciate your PR and your patience as we work to improve our developer experience, and welcome you to re-open your changes against the new location. Please reach out if you have any questions!
